src/app/rpmostree-compose-builtin-tree.c

```diff
@@ -496,52 +490,88 @@ rpm_ostree_compose_context_new (const char    *treefile_pathstr,
 
   if (opt_unified_core)
     {
+      /* Unified mode works very differently. We ignore --workdir because we want to be sure
+       * that we're going to get hardlinks. The only way to be sure of this is to place the
+       * workdir underneath the pkgcache repo. */
+
       if (opt_workdir)
         g_printerr ("note: --workdir is ignored for --ex-unified-core\n");
 
-      /* For unified core, our workdir must be underneath the repo tmp/
-       * in order to use hardlinks.  We also really want a bare-user repo.
-       * We hard require that for now, but down the line we may automatically
-       * do a pull-local from the bare-user repo to the archive.
+      /* We also really want a bare-user repo. We hard require that for now, but down the
+       * line we may automatically do a pull-local from the bare-user repo to the archive.
        */
       if (ostree_repo_get_mode (self->repo) != OSTREE_REPO_MODE_BARE_USER)
         return glnx_throw (error, "--ex-unified-core requires a bare-user repository");
-      if (!glnx_mkdtempat (ostree_repo_get_dfd (self->repo), "tmp/rpm-ostree-compose.XXXXXX", 0700,
-                           &self->workdir_tmp, error))
-        return FALSE;
-      /* Note special handling of this aliasing in _finalize() */
-      self->workdir_dfd = self->workdir_tmp.fd;
 
-    }
-  else if (!opt_workdir)
-    {
-      if (!glnx_mkdtempat (AT_FDCWD, "/var/tmp/rpm-ostree.XXXXXX", 0700, &self->workdir_tmp, error))
+      if (opt_cachedir)
+        {
+          if (!glnx_opendirat (AT_FDCWD, opt_cachedir, TRUE, &self->cachedir_dfd, error))
+            {
+              g_prefix_error (error, "Opening cachedir '%s': ", opt_cachedir);
+              return FALSE;
+            }
+
+          /* Put workdir beneath cachedir, which is where the pkgcache repo also is */
+          if (!glnx_mkdtempat (self->cachedir_dfd, "rpm-ostree-compose.XXXXXX", 0700,
+                               &self->workdir_tmp, error))
+            return FALSE;
+        }
+      else
+        {
+          /* Put cachedir under the target repo: makes things more efficient if it's
+           * bare-user, and otherwise just restricts IO to within the same fs. If for
+           * whatever reason users don't want to run the compose there (e.g. weird
+           * filesystems that aren't fully POSIX compliant), they can just use --cachedir.
```

Comment:
I wonder though if this is the right thing for NFS.  We have `repo_is_on_netfs()` and could use it here.  But eh...

Reply:
Seems reasonable. Fixup!